client/landing/stepper/index.tsx

```diff
@@ -149,3 +152,7 @@ window.AppBoot = async () => {
 		document.getElementById( 'wpcom' )
 	);
 };
+
+if ( module.hot ) {
+	module.hot.accept();
+}
```

Comment:
This explicit `module.hot.accept()` shouldn't be needed. Or did you have to add it in order to make the Stepper landing page work?

The `react-refresh` webpack runtime adds code to the end of every module, and this code calls `module.hot.accept()` if the module exports only React components. Only for such modules, `react-refresh` can handle the hot update and replace the module at runtime without causing any errors.

For other modules, hot updates won't be accepted. Like when you modify a reducer or something like that. Then we can't guarantee that the new code will be really used. References to old code can still be in place.

This is how `react-refresh` works: it supports hot updates only for React components and nothing else.

If you want to support hot reloading for the Stepper's top-level UI, the best approach is to extract it to a separate file and then import it:
```js
import Stepper from './component';
render( <Stepper flow={ flow } />, document.getElementById( 'wpcom' ) );
```

Reply:
I think it is needed. I tested this branch by removing the explicit reloads:

```diff
diff --git a/client/boot/app.js b/client/boot/app.js
index 58f23397f7..dcf82c105e 100644
--- a/client/boot/app.js
+++ b/client/boot/app.js
@@ -8,7 +8,3 @@ import 'calypso/assets/stylesheets/style.scss';
 window.AppBoot = () => {
        bootApp( 'Calypso' );
 };
-
-if ( module.hot ) {
-       module.hot.accept();
-}
diff --git a/client/landing/stepper/index.tsx b/client/landing/stepper/index.tsx
index 0835a41c27..0df1207135 100644
--- a/client/landing/stepper/index.tsx
+++ b/client/landing/stepper/index.tsx
@@ -152,7 +152,3 @@ window.AppBoot = async () => {
                document.getElementById( 'wpcom' )
        );
 };
-
-if ( module.hot ) {
-       module.hot.accept();
-}
```

Then tried basic hot-reloading by making changes to the html provided by `client/my-sites/customer-home/main.jsx` and `client/landing/stepper/declarative-flow/internals/steps-repository/store-profiler/index.tsx`. (nothing in the data layer). I found that most of the time, the updates didn't work:

Screenshots from a session where it didn't work:

![2023-05-12_10-05](https://github.com/Automattic/wp-calypso/assets/937354/93fd6e68-6ffe-42c6-afbd-bb9bc95f55fb)
![2023-05-12_10-05_1](https://github.com/Automattic/wp-calypso/assets/937354/74e4ee0e-dc8d-4a5e-915f-b558c70ab6bd)

I'm not sure what's happening though, because I think it did work with them removed at least once. However, I did just restart calypso 5 or 6 times and can't seem to get any hot reloads working with these removed anymore.

Reply:
It works for with the `accept` calls removed when I change only the JSX markup in `customer-home/main.jsx`. Then the dev server creates updates for the `home` and `runtime` assets:
```
webpack building...
webpack built be80d0323231fa6ba8ad in 4178ms
assets by status 167 MiB [cached] 2126 assets
assets by path . 1.12 MiB
  assets by info 21.3 KiB [immutable]
    asset home.3f0d1f8c4c2e3e97aa60.hot-update.js 20.4 KiB [emitted] [immutable] [hmr] (name: home)
    asset runtime.3f0d1f8c4c2e3e97aa60.hot-update.js 859 bytes [emitted] [immutable] [hmr] (name: runtime)
    asset runtime.3f0d1f8c4c2e3e97aa60.hot-update.json 38 bytes [emitted] [immutable] [hmr]
  asset home.js 1.02 MiB [emitted] (name: home)
  asset runtime.js 75.5 KiB [emitted] (name: runtime)
cached modules 51.6 MiB (javascript) 18 MiB (asset) 12.7 MiB (css/mini-extract) [cached] 15736 modules
runtime modules 56.2 KiB 24 modules
./my-sites/customer-home/main.jsx 9.74 KiB [built] [code generated]
webpack 5.68.0 compiled successfully in 4178 ms

Ready! All assets are re-compiled. Have fun!
```
and the browser accepts the update:
```
[webpack] hot updated 1 modules:
webpack-client.ts:97 ./my-sites/customer-home/main.jsx
```
because the `main.jsx` module has `react-refresh` code, added by the Babel and webpack plugins, that accepts the update if the React exports are compatible.

In your screenshots, however, I see there is also an update for the `sections` module, which is not accepted because its exports aren't React components:
```
[webpack] hot update failed: Error: Aborted because ./sections.is is not accepted
Update propagation: ./sections.js -> ./sections-middleware.js -> ./boot/common. js -> ./boot/app.js
```
Here webpack also asks the parent modules if they `accept` and that's why `module.hot.accept()` in the `./boot/app.js` module works.

I'm wondering where this `sections` update is coming from. How are the old and new modules different? What does webpack report on the server, in the terminal console, when building this update?

Anyway, I think you can merge this PR even with the `module.hot.accept()` calls, they are harmless. Once we figure out what's going on with the `sections` update we can fix that in a followup.

Reply:
Adding an unconditional `module.hot.accept()` call to the entrypoint module also has the effect that _every_ hot update will be accepted, unless a child module, i.e., the `react-refresh` runtime, explicitly rejected it. That's not what we want, we want to accept only updates that can be safely applied.